src/core/Pipelining.ts

```diff
@@ -0,0 +1,181 @@
+     *  by default.
+     * @throws {AssertionError} if the depth is smaller than 1.
+     * @throws {AssertionError} if the load generator is null.
+     */
+    constructor(depth: number, loadGenerator: () => Promise<E>, storeResults: boolean = false) {
+        assertPositive(depth, 'depth should be positive');
+        assertNotNull(loadGenerator, 'load generator cannot be null');
+        this.depth = depth;
+        this.loadGenerator = loadGenerator;
+        if (storeResults) {
+            this.next = this.nextWithResults;
+        } else {
+            this.next = this.nextWithoutResults;
+        }
+    }
+
+    /**
+     * If the Pipelining is constructed to store the results, results will be
+     * returned with order the requests were done when all of the requests
+     * are completed.
+     *
+     * If not, it will return void when all of the requests are completed.
+     *
+     * @throws {Error} if any of the requests fails
+     * @returns array of the results or void
+     */
+    run(): Promise<void | E[]> {
+        this.reset();
+        return new Promise<void | E[]>((resolve, reject) => {
+            this.resolve = resolve;
+            this.reject = reject;
+
+            for (let i = 0; i < this.depth; i++) {
+                this.next();
+                if (this.done) {
+                    break;
+                }
+            }
+        });
+    }
+
+    /**
+     * Sets a new load generator for this Pipelining.
+     *
+     * @param loadGenerator a generator-like function that returns a promise
+     *  or null if the generator is exhausted.
+     *
+     * @throws {AssertionError} if the load generator is null.
+     */
+    setLoadGenerator(loadGenerator: () => Promise<E>): void {
+        assertNotNull(loadGenerator, 'load generator cannot be null');
+        this.loadGenerator = loadGenerator;
+    }
+
+    private nextWithoutResults(): void {
+        if (this.rejected) {
+            return;
+        }
+
+        const promise = this.loadGenerator();
+        if (promise == null) {
+            this.done = true;
+            if (this.resolvingCount === 0) {
+                this.resolve();
+            }
+            return;
+        }
+
+        this.resolvingCount++;
+        promise.then(() => {
+           this.resolvingCount--;
+           this.next();
+        }).catch((err) => {
+            this.rejected = true;
+            this.reject(err);
+        });
+    }
+
+    private nextWithResults(): void {
+        if (this.rejected) {
+            return;
+        }
+
+        const promise = this.loadGenerator();
+        const i = this.index++;
+
+        if (promise == null) {
+            this.done = true;
+            if (this.resolvingCount === 0) {
+                this.resolve(this.results);
+            }
+            return;
+        }
+
+        this.resolvingCount++;
+        promise.then((value) => {
+           this.results[i] = value;
```

Comment:
`nextWithResults` and `nextWithoutResults` are the same except this line. Can you reduce code duplication more?